native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeObjCNameChecker.kt

```diff
@@ -0,0 +1,144 @@
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_OBJC_NAME_FIRST_CHAR.on(reportLocation, invalidFirstChars.joinToString("")))
+        }
+        val invalidNameChars = objCName.name?.toSet()?.subtract(validChars) ?: emptySet()
+        val invalidSwiftNameChars = objCName.swiftName?.toSet()?.subtract(validChars) ?: emptySet()
+        val invalidChars = invalidNameChars + invalidSwiftNameChars
+        if (invalidChars.isNotEmpty()) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_OBJC_NAME_CHARS.on(reportLocation, invalidChars.joinToString("")))
+        }
+        if (objCName.exact && descriptor !is ClassDescriptor) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INAPPLICABLE_EXACT_OBJC_NAME.on(reportLocation))
+        }
+        if (objCName.exact && objCName.name == null) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.MISSING_EXACT_OBJC_NAME.on(reportLocation))
+        }
+    }
+
+    private fun checkOverrides(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        if (descriptor !is CallableMemberDescriptor || descriptor.overriddenDescriptors.isEmpty()) return
+        val objCNames = descriptor.overriddenDescriptors.map { it.getFirstBaseDescriptor().getObjCNames() }
+        if (!objCNames.allNamesEquals()) {
+            val containingDeclarations = descriptor.overriddenDescriptors.map { it.containingDeclaration }
+            context.trace.report(ErrorsNative.INCOMPATIBLE_OBJC_NAME_OVERRIDE.on(declaration, descriptor, containingDeclarations))
+        }
+    }
+
+    private fun checkFakeOverrides(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        if (descriptor !is ClassDescriptor) return
+        descriptor.defaultType.memberScope
+            .getContributedDescriptors(DescriptorKindFilter.ALL, MemberScope.Companion.ALL_NAME_FILTER)
+            .forEach {
+                if (it !is CallableMemberDescriptor || it.kind.isReal) return@forEach
+                checkOverrides(declaration, it, context)
+            }
+    }
+
+    private fun CallableMemberDescriptor.getFirstBaseDescriptor(): CallableMemberDescriptor =
+        if (overriddenDescriptors.isEmpty()) this else overriddenDescriptors.first().getFirstBaseDescriptor()
+
+    private class ObjCName(
+        val annotation: AnnotationDescriptor
+    ) {
+        val name: String? = annotation.argumentValue("name")?.value?.cast<String>()?.takeIf { it.isNotBlank() }
```

Comment:
I believe all three of these `cast` usages will throw `ClassCastException`'s if the usage of `ObjCName` is with incorrect types: `@ObjCName(32)`. We should probably check the types instead and report the type mismatch as a diagnostics?

The type error is probably already reported, so when checking we can probably just do nothing if a name, swiftName or exact parameter with the wrong type is present?

Reply:
> I believe all three of these cast usages will throw ClassCastException's if the usage of ObjCName is with incorrect types: @ObjCName(32). We should probably check the types instead and report the type mismatch as a diagnostics?

Shouldn't the annotation be resolved and valid at this point? Meaning the value of `name` is known to be a `String`?

Reply:
The diagnostics checkers are run on invalid code as well; since they are really there to tell you about problem in the code both in the IDE and on the command-line. I just tried this out by compiling a program with the standalone command-line konanc where there is a type incorrect use of `@ObjCName` and it leads to an uncaught `ClassCastException` right now.